src/Equinox/Decider.fs

```diff
@@ -290,3 +288,34 @@ and internal SyncContext<'state> =
             member _.Version = token.version
             member _.StreamEventBytes = match token.streamBytes with -1L -> ValueNone | b -> ValueSome b
             member _.CreateMemento() = token, state }
+
+[<Struct>]
+type Target =
+    { categoryName : string; streamId : string }
+    static member Render(categoryName, streamId) = String.Concat(categoryName, '-', streamId)
+    override x.ToString() = Target.Render(x.categoryName, x.streamId)
+
+module Target =
+    module Internal =
+        /// Throws if a candidate category includes a '-', is null, or is empty
+        let inline validateCategory (rawCategory : string) =
+            if rawCategory |> String.IsNullOrEmpty then invalidArg "rawCategory" "may not be null or empty"
+            if rawCategory.IndexOf '-' <> -1 then invalidArg "rawCategory" "may not contain embedded '-' symbols"
+        /// Throws if a candidate id element includes a '_', is null, or is empty
+        let inline validateElement (rawElement : string) =
+            if rawElement |> String.IsNullOrEmpty then invalidArg "rawElement" "may not contain null or empty components"
+            if rawElement.IndexOf '_' <> -1 then invalidArg "rawElement" "may not contain embedded '_' symbols"
+        /// Low level helper used to gate ingestion from a canonical form
+        /// Does NOT validate the {streamId} portion wrt numbers of embedded `_` or `-` chars
+        let create categoryName streamId =
+            validateCategory categoryName
+            { categoryName = categoryName; streamId = streamId }
+        let toString (x : Target) = string x
+
+    let private streamIdOfElement f id = let t = f id in Internal.validateElement t; t
```

Comment:
This function desperately wants to escape to the plains of its sacred homeland, where the OCamls roam. 😂

Reply:
fine! (I tried to unroll it so many times but hated it)